lib/rules/no-loop-func.js

```diff
@@ -186,13 +198,49 @@ module.exports = {
             }
 
             const references = sourceCode.getScope(node).through;
-            const unsafeRefs = references.filter(r => r.resolved && !isSafe(loopNode, r)).map(r => r.identifier.name);
+            const unsafeRefs = references.filter(r => r.resolved && !isSafe(loopNode, r));
+            const unsafeRefsName = unsafeRefs.map(r => r.identifier.name);
+
+            // Check if the function is not asynchronous or a generator function
+            if (!(node.async || node.generator)) {
+                const parentNode = node.parent;
+                const isFunctionExpression = node.type === "FunctionExpression";
+                const ArrowFunctionExpression = node.type === "ArrowFunctionExpression";
+
+                if (parentNode.type === "CallExpression" && (isFunctionExpression || ArrowFunctionExpression)) {
+
+                    // Check if the function is referenced elsewhere in the code
+                    const isFunctionReferenced = isFunctionExpression && node.id ? references.some(r => r.identifier.name === node.id.name) : false;
+
+                    // Check if there are unsafe references used within non-immediately-invoked nested functions
+                    const hasUnsafeRefsInNonIIFE = unsafeRefs.some(r => {
+                        let refFunctionNode = r.from.variableScope.block;
+
+                        if (node === refFunctionNode) {
+                            return !isIIFE(refFunctionNode);
+                        }
+
+                        while (node !== refFunctionNode) {
+                            if (!isIIFE(refFunctionNode)) {
+                                return true;
+                            }
+                            refFunctionNode = refFunctionNode.parent;
```

Comment:
We don't need to check if every ancestor node is an IIFE, just if every separate execution context (e.g., an inner function) is immediately executed. If the innermost is `r.from.variableScope`, the next upper would be `r.from.variableScope.upper.variableScope` and so on.

Reply:
Updated, thanks!